### ISSUE TITLE
Vulkan: MeshInstance3D looses Shader Instance Uniform Color when loaded with a new mesh

### ISSUE THREAD

**Issue by JoanPotatoes2021** [godotengine/godot#58113]
### Godot version

v4.0.alpha2.official [79077e6c1]

### System information

Windows 10

### Issue description

Changing the mesh of a MeshInstance3D breaks any colorUniform from the shader assigned with Instance Uniform qualifier. While we can still print it's color value.

![ShaderUniformBreakMeshLoad](https://user-images.githubusercontent.com/88800565/153940710-c8b2a586-41a8-4895-b536-d727612e1b7d.gif)

Editing texturesUniform parameters not instance qualified works fine, but instance uniform colors seems to go black and not respond to any further parameter edits.

### Steps to reproduce

1 Create a shader material and assign a color instance uniform to it.
2 Change the MeshInstance3D by doing load(another_mesh_path)
3 Color instance uniforms are perpetually black.

### Minimal reproduction project

[LoadingMeshBreaksInstanceUniform.zip](https://github.com/godotengine/godot/files/8063723/LoadingMeshBreaksInstanceUniform.zip)

**Comment by GNSS-Stylist** [comment]
Stumbled upon this same issue and almost created a new issue, but found this so adding my findings here instead:
Video:

https://user-images.githubusercontent.com/50438441/218575725-cc8cde75-ed18-4ac0-a68f-f3b5f031e8ec.mp4

Funnily OBS didn't seem to grab the popups when selecting a new mesh nor when changing colors...

First I also stumbled upon this issue when working with MeshInstance3D, but this also affects CSGShape3Ds, so the problem might be inherited from GeometryInstance3D(?) It also seems to affect also other types of instance uniforms (or at least floats) in addition to color.

It doesn't matter if changing the geometry is done in inspector, running the project or tool script. In MRP I tried several different workarounds (like storing/restoring a duplicated material and the value of the instance uniform) to no avail. I left these with some other tests into MRP so they can be used in tests if found useful.

Last issue mentioned in the video (values of instance uniforms changing to defaults when shader compilation fails) seems to actually be duplicate of https://github.com/godotengine/godot/issues/64057

My MRP:
[MeshInstance3DUpdateBug.zip](https://github.com/godotengine/godot/files/10726351/MeshInstance3DUpdateBug.zip)

Godot version: v4.0.rc.custom_build [4a2c2170b]
